flake.nix

```diff
@@ -74,17 +66,29 @@
       name = "tailscale";
 
       src = ./.;
-      vendorSha256 = fileContents ./go.mod.sri;
-      nativeBuildInputs = pkgs.lib.optionals pkgs.stdenv.isLinux [ pkgs.makeWrapper pkgs.git ];
+      vendorSha256 = pkgs.lib.fileContents ./go.mod.sri;
+      nativeBuildInputs = pkgs.lib.optionals pkgs.stdenv.isLinux [ pkgs.makeWrapper ];
       ldflags = ["-X tailscale.com/version.GitCommit=${tailscaleRev}"];
       CGO_ENABLED = 0;
       subPackages = [ "cmd/tailscale" "cmd/tailscaled" ];
       doCheck = false;
+
+      # NOTE: We strip the ${PORT} and $FLAGS because they are unset in the
+      # environment and cause issues (specifically the unset PORT). At some
+      # point, there should be a NixOS module that allows configuration of these
+      # things, but for now, we hardcode a default of port 0 (which causes the
+      # kernel to delegate a port).
       postInstall = pkgs.lib.optionalString pkgs.stdenv.isLinux ''
         wrapProgram $out/bin/tailscaled --prefix PATH : ${pkgs.lib.makeBinPath [ pkgs.iproute2 pkgs.iptables pkgs.getent pkgs.shadow ]}
         wrapProgram $out/bin/tailscale --suffix PATH : ${pkgs.lib.makeBinPath [ pkgs.procps ]}
 
-        sed -i -e "s#/usr/sbin#$out/bin#" -e "/^EnvironmentFile/d" ./cmd/tailscaled/tailscaled.service
+        sed -i \
+          -e "s#/usr/sbin#$out/bin#" \
+          -e "/^EnvironmentFile/d" \
+          -e 's/''${PORT}/0/' \
```

Comment:
Could you set the default to 41641 instead? That will make tailscaled try to get a static port for wireguard comms (which is useful for debugging, opening the right port in the firewall, etc.), but it still falls back to a random port if 41641 is unavailable.

Reply:
Thanks for your input, @danderson. It begs why we don't default to 41641 in the CLI.

Reply:
I'll force-push away the commit changing to port 0 :+1:

Reply:
Indeed. The answer is an unfortunate set of historical reasons where the current defaults for the daemon and CLI evolved organically over time :). We also spent a bunch of time trying to work around different classes of bugs, with different tradeoffs (e.g. make NAT traversal worse on average by roaming the binding port, but avoid problems with buggy NAT devices that incorrectly handle multiple clients using the same port). The default in tailscaled also influences the behavior of the Windows client and some macOS clients, which have their own set of tradeoffs.

This is a fairly big can of worms to fix in a principled way. We should do that, but in the meantime our choices are: do we default to the settings we configure for systemd linux distros (41641, per cmd/tailscaled/tailscaled.defaults), or do we default to whatever the daemon does by default (0) ? My inclination would be to stick to the defaults we picked for systemd distros in general, until we go back and clean this up throughout the codebase, and accounting for the issues with all clients.

(really this needs a major magicsock change to support the use of multiple concurrent sockets, which has been on the list for a long time but is also non-trivial, sadly)

Reply:
Doh, I missed the earlier conversation where shayne asked you to change it in the other direction. Sorry for the contradictory instructions! Both settings definitely have their merits here, and if I'd seen the previous conversation I would probably have just said "eh, okay that's fine, we can defer this until we fix it overall".

I still think on balance sticking to the setting we ship to other systemd distros is preferable, given that I forced you to do extra work already. Sorry again, and thanks :pray:

Reply:
No worries whatsoever! I did think the contradiction was funny, but I think both are good choices (and unifying at least the Linuxes on 41641 seems like a good enough choice).